pkg/controllers/statusaggregator/plugins/job.go

```diff
@@ -0,0 +1,163 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/common"
+)
+
+type JobPlugin struct{}
+
+func NewJobPlugin() *JobPlugin {
+	return &JobPlugin{}
+}
+
+func (receiver *JobPlugin) AggregateStatuses(
+	ctx context.Context,
+	sourceObject, fedObject *unstructured.Unstructured,
+	clusterObjs map[string]interface{},
+	clusterObjsUpToDate bool,
+) (*unstructured.Unstructured, bool, error) {
+	logger := klog.FromContext(ctx).WithValues("status-aggregator-plugin", "jobs")
+
+	needUpdate := false
+	aggregatedStatus := &batchv1.JobStatus{}
+	var completionTime *metav1.Time
+	var completionCount int
+	var jobCompleted, jobFailed []string
+
+	for clusterName, clusterObj := range clusterObjs {
+		logger := klog.FromContext(ctx).WithValues("cluster-name", clusterName)
+
+		utd := clusterObj.(*unstructured.Unstructured)
+		// For status of job
+		var found bool
+		status, found, err := unstructured.NestedMap(utd.Object, common.StatusField)
+		if err != nil || !found {
+			logger.Error(err, "Failed to get status of cluster object")
+			return nil, false, err
+		}
+
+		if status == nil {
+			continue
+		}
+
+		jobStatus := &batchv1.JobStatus{}
+		if err = runtime.DefaultUnstructuredConverter.FromUnstructured(status, jobStatus); err != nil {
+			logger.Error(err, "Failed to convert the status of cluster object")
+			return nil, false, err
+		}
+
+		if aggregatedStatus.StartTime.IsZero() || jobStatus.StartTime.Before(aggregatedStatus.StartTime) {
+			aggregatedStatus.StartTime = jobStatus.StartTime
+		}
+		if !jobStatus.CompletionTime.IsZero() {
+			completionCount++
+			jobCompleted = append(jobCompleted, clusterName)
+			if completionTime.IsZero() || completionTime.Before(jobStatus.CompletionTime) {
+				completionTime = jobStatus.CompletionTime
+			}
+		} else if IsJobFinishedWithFailed(jobStatus) {
+			completionCount++
+			jobFailed = append(jobFailed, clusterName)
+		}
+
+		aggregatedStatus.Active += jobStatus.Active
+		aggregatedStatus.Succeeded += jobStatus.Succeeded
+		aggregatedStatus.Failed += jobStatus.Failed
+	}
+	if completionCount > 0 && completionCount == len(clusterObjs) {
```

Comment:
If there is a `Failed` job, should we wait until all jobs are finished before marking the host job as `Failed`? @mrlihanbo @limhawjia

Reply:
Semantically, I think a single failed job should mean that the host job has failed. However, setting it immediately would cause us to lose information about the remaining jobs if we choose not to update finished jobs (as described below)... I'm still not sure what's the best approach